Fix build issues and create Travis job matrix

CamStan opened the pull request:
### Description
<!--- Describe your changes in detail -->

Adjust UnifyFS to build and test properly with or without GOTCHA
- Adjust examples/src/Makefile.am to not build HDF5 examples when
  GOTCHA is unavailable
- Fix configure.ac -wrap test that fails when providing additional
  LDFLAGS during build
- Adjust t/Makefile.am to omit GOTCHA unit tests when building without
  GOTCHA
- Fix indentation issues in t/Makefile.am causing conditionals to be
  missed and always treated as "false" when building with GOTCHA

Fix build with Fortran and without GOTCHA
- Adjust client/src/Makefile.am and examples/src/Makefile.am to embed
  the `HAVE_FORTRAN` section within the `HAVE_GOTCHA` section
- Set configure to fail when attempting to build with Fortran enabled
  but GOTCHA wasn't found
- The latter two enforces the requirement of GOTCHA when enabling
  Fortran support

Add spath config option
- Create a simple --with-spath config option so users can configure using
  only "--with" options (need for distcheck), only CPPFLAGS/LDFLAGS, or both
- Skip unit tests that require spath when missing spath dependency

Note: Having "--with" options allows continued use of `distcheck` in
Travis, as `DISTCHECK_CONFIGURE_FLAGS` doesn't work with quotes
inside of options (e.g., multiple CPPFLAGS/LDFLAGS)

Set up jobs matrix in .travis.yml to test on different compilers (can see it [here](https://travis-ci.com/github/LLNL/UnifyFS/builds/219825383))
- Create CI jobs to build and test on GCC 4.9, 7.5, 9.3 and 10.1
- Pull syntax checker out into separate CI job so it only runs once
  and passes/fails independently of the different compiler builds
- Fix errors on new compilers
  - Fixes Spack issue that was patched for latest release (https://github.com/spack/spack/issues/21983)
  - Fixes issue #609
- Adjust configure to take GOTCHA and SPATH paths (or CPPFLAGS/LDFLAGS)
  to account for recent Spack change
  - Couldn't set Travis to always use latest Spack release as it
    still has much older version of the packages (e.g.,`margo@0.4.3`)

Update docs
- Update build docs
- Note GOTCHA requirement when enabling Fortran
- Add spath option to config sections
- Warn users of MPI-IO and HDF5 support loss when not using GOTCHA
- Update Margo dependency to v0.9.1
- Update Travis CI urls in README

Additional
- Set min `automake` version to 1.15

### Updated
Update bootstrap.sh
- Update urls for migrated repos (mochi-margo and bmi)
- Set Margo version to v0.9.1
- Remove old Mercury cmake options
- Add `-std=gnu99` cflags to margo configure if gcc is older than v5.x
- Update exports printed at end

Update .gitlab-ci.yml
- change loaded dependencies in `before_script`
- move envars only needed for `build` stage out of global `before_script`
- update `configure` line to account for build fixes
- unbuffer integration testing output in Gitlab CI web interface to avoid timeouts

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
### Updated
- Fixes: #608
- Fixes: #609 

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->
- Locally using all combinations of `--enable/disable-fortran`,
`--with/without-gotcha`, and `--with/without-spath`.
- On multiple GCC compiler versions in our new Travis jobs matrix

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Testing (addition of new tests or update to current tests)
- [x] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the UnifyFS code style requirements.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] All commit messages are properly formatted.

CamStan commented:
Feel free to review this, but I may add bootstrap fixes to this PR as well as some more doc updates as I get our spackage updated and working.

CamStan commented on client/src/unifyfs-sysio.c:
```diff
@@ -246,7 +246,7 @@ static char* _getcwd_impl(char* path, size_t size)
         buf = (char*) malloc(len);
         if (buf != NULL) {
-            strncpy(buf, unifyfs_cwd, len);
+            memcpy(buf, unifyfs_cwd, strlen(unifyfs_cwd));
```
This was the fix used to patch our Spack package in https://github.com/spack/spack/pull/22066.
Want to check that this is the solution we want to use here.

MichaelBrim replied:
It is fine to use memcpy, but you need to use `strlen(string) + 1` for the length to make sure the NUL character is included. `strncpy` makes sure the string is NUL-terminated implicitly.

CamStan replied:
Updated.

adammoody replied:
Yep, that looks good.  Should we use ``strlcpy`` here to be consistent?

It's bigger than this PR, but grep turns up a bunch of other ``strncpy`` hits.  Maybe it's worth doing a full sweep of the code base to review a full replacement of ``strncpy`` with ``strlcpy``.

CamStan replied:
I'm happy to change this one to `strlcpy` for consistency and we can tackle the others in another PR. These were the only two the new compilers were currently complaining about.

CamStan commented on client/src/unifyfs.c:
```diff
@@ -924,7 +924,8 @@ int unifyfs_fid_create_file(const char* path)
 
     /* copy file name into slot */
-    strncpy((void*)&unifyfs_filelist[fid].filename, path, UNIFYFS_MAX_FILENAME);
+    strncpy((void*)&unifyfs_filelist[fid].filename, path,
+             UNIFYFS_MAX_FILENAME - 1);
```
This is to resolve the build error that showed up on GCC 10.1 which I believe is the same error in #609.
Want to check that this is the solution we want to use here.

MichaelBrim replied:
In other places where newer GCC was complaining about buffer sizes, Tony switched from `strncpy` to `strlcpy` implemented in `common/src/unifyfs_misc.[ch]`

CamStan replied:
Switched to `strlcpy`.

CamStan commented on docs/build.rst:
```diff
@@ -243,6 +255,15 @@ provided by either PMI2 or PMIx. To enable this support, pass either
 the ``--enable-pmi`` or ``--enable-pmix`` option to configure. Without
 PMI support, a distributed file system accessible to all servers is required.
 
+SPATH
+******
+
+The spath library can be optionally used to normalize relative paths (e.g., ones
+containing ".", "..", and extra or trailing "/") and enable the support of using
+relative paths within an application. To enable, use the ``--with-spath``
+configure option or provide the appropriate ``CPPFLAGS`` and ``LDFLAGS`` at
+configure time.
```
Not sure I captured this properly. @adammoody, let me know if you have any suggestions.

adammoody replied:
Yeah, that looks good.  Thanks.

adammoody commented:
Just by reading your description alone, I can tell this is really great and detailed work, @CamStan .  Thanks a ton!

I'll review the code in more detail tomorrow.

MichaelBrim commented on docs/build.rst:
```diff
@@ -103,40 +102,55 @@ the UnifyFS dependencies can then be installed.
+
+.. code-block:: Bash
+
+    $ ./configure --prefix=/path/to/install CPPFLAGS="-I${gotcha_install}include -I{spath_install}/include" LDFLAGS="-L${gotcha_install}/lib64 -L${spath_install}/lib64
```
@CamStan you're missing a `/` in gotcha include

CamStan replied:
Yes, yes I am.
Added it.

adammoody reviewed (approved):
Thanks, @CamStan !  Looks good to me.  I'll wait for @MichaelBrim to also sign off before we merge.

CamStan commented:
Updated the `bootstrap.sh` script but didn't need much to work for me.

Shouldn't have anything more coming for this PR. Feel free to look over this again.

CamStan commented:
I lied... I also updated _.gitlab-ci.yml_ to account for changes in this PR in order to get the CI tests running again. I promise that's the last thing. 🤞 

@MichaelBrim and @adammoody, this is ready for a final review. I made Mike's suggested changes and added changes to the _bootstrap.sh_ script and _.gitlab-ci.yml_.

MichaelBrim reviewed (approved):
Everything looks good to me.
